File: dsa-problems/GraphsBasics/solutions/python/GRB-009-bipartite-check-bfs.py
import sys
from collections import deque

def bipartite_colors(n: int, adj: list[list[int]]):
    colors = [-1] * n
    
    for i in range(n):
        if colors[i] == -1:
            queue = deque([i])
            colors[i] = 0
            
            while queue:
                u = queue.popleft()
                for v in adj[u]:
                    if colors[v] == -1:
                        colors[v] = 1 - colors[u]
                        queue.append(v)
                    elif colors[v] == colors[u]:
                        return None
                        
    return colors

def main():
    input = sys.stdin.read
    data = input().split()
    if not data:
        return
    
    iterator = iter(data)
    try:
        n = int(next(iterator))
        m = int(next(iterator))
        adj = [[] for _ in range(n)]
        for _ in range(m):
            u = int(next(iterator))
            v = int(next(iterator))
            adj[u].append(v)
            adj[v].append(u)
            
        colors = bipartite_colors(n, adj)
        if colors is None:
            print("0")
        else:
            print("1")
    except StopIteration:
        pass

if __name__ == "__main__":
    main()

File: dsa-problems/GraphsBasics/solutions/javascript/GRB-009-bipartite-check-bfs.js
const readline = require("readline");

class Solution {
  bipartiteColors(n, adj) {
    const colors = new Int32Array(n).fill(-1);

    for (let i = 0; i < n; i++) {
      if (colors[i] === -1) {
        const queue = [i];
        colors[i] = 0;
        let head = 0;
        
        while (head < queue.length) {
          const u = queue[head++];
          
          for (const v of adj[u]) {
            if (colors[v] === -1) {
              colors[v] = 1 - colors[u];
              queue.push(v);
            } else if (colors[v] === colors[u]) {
              return null;
            }
          }
        }
      }
    }
    return colors;
  }
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => data.push(...line.trim().split(/\s+/)));
rl.on("close", () => {
  if (data.length === 0) return;
  
  let idx = 0;
  const n = parseInt(data[idx++], 10);
  const m = parseInt(data[idx++], 10);
  
  const adj = Array.from({ length: n }, () => []);
  for (let i = 0; i < m; i++) {
    const u = parseInt(data[idx++], 10);
    const v = parseInt(data[idx++], 10);
    adj[u].push(v);
    adj[v].push(u);
  }

  const solution = new Solution();
  const colors = solution.bipartiteColors(n, adj);
  
  if (colors === null) {
    console.log("false");
  } else {
    console.log("true");
    console.log(colors.join(" "));
  }
});
